content/blog/datafusion-custom-parquet-index.md

```diff
@@ -0,0 +1,232 @@
+
+### Binary Layout in the Parquet File
+
+```text
+                   ┌──────────────────────┐
+                   │┌───────────────────┐ │
+                   ││     DataPage      │ │
+                   │└───────────────────┘ │
+  Standard Parquet │┌───────────────────┐ │
+  Data Pages       ││     DataPage      │ │
+                   │└───────────────────┘ │
+                   │        ...           │
+                   │┌───────────────────┐ │
+                   ││     DataPage      │ │
+                   │└───────────────────┘ │
+                   │┏━━━━━━━━━━━━━━━━━━━┓ │
+ Non‑standard      │┃                   ┃ │
+ index (ignored by │┃ Custom Binary     ┃ │
+ other Parquet     │┃ Index (Distinct)  ┃◀┼─ ─ ─
+ readers)          │┃                   ┃ │   │  Inline after
+                   │┗━━━━━━━━━━━━━━━━━━━┛ │   │  data pages
+                   │┏━━━━━━━━━━━━━━━━━━━┓ │   │
+ Standard Parquet  │┃    Page Index     ┃ │   │
+ Page Index        │┗───────────────────┛ │   │  Footer
+                   │╔═══════════════════╗ │   │  metadata
+                   │║ Parquet Footer w/ ║ │◀──┼─ ─ ─ key/value
+                   │║     Metadata      ║ │      map contains
+                   │║ (Thrift Encoded)  ║ │      offset
+                   │╚═══════════════════╝ │
+                   └──────────────────────┘
+```
+
+Embedding a custom distinct‑value index inside Parquet files empowers DataFusion to skip non‑matching files with zero additional files to manage.
+
+---
+
+## Example Implementation Walkthrough
+
+### Serializing the Index in Rust
+
+```rust
+/// Magic bytes to identify our custom index format
+const INDEX_MAGIC: &[u8] = b"IDX1";
+
+/// Serialize the distinct index using ArrowWriter to account for byte offsets
+fn serialize_index<W: Write + Send>(
+    index: &DistinctIndex,
+    writer: &mut ArrowWriter<W>,
+) -> Result<()> {
+    // Get current byte offset (end of data pages)
+    let offset = writer.bytes_written();
+
+    // Build newline‑separated UTF‑8 payload
+    let serialized = index.inner.iter().cloned().collect::<Vec<_>>().join("\n");
+    let bytes = serialized.into_bytes();
+
+    // Write magic + length
+    writer.write_all(INDEX_MAGIC)?;
+    writer.write_all(&(bytes.len() as u64).to_le_bytes())?;
+
+    // Write index payload
+    writer.write_all(&bytes)?;
+
+    // Append offset metadata to footer
+    writer.append_key_value_metadata(
+        KeyValue::new(
+            "distinct_index_offset".to_string(),
+            offset.to_string(),
+        ),
+    );
+    Ok(())
+}
+```
+
+### Reading the Index
+
+On the read path, we:
+
+1. Open the Parquet footer and extract `distinct_index_offset`.
+2. Seek to that offset in the file.
+3. Read and validate `IDX1` magic.
+4. Read the 8‑byte length and then the payload.
+5. Reconstruct `DistinctIndex` from newline‑delimited strings.
```

Comment:
I suggest cuting out the details of reading the index so

1. Open the Parquet footer and extract `distinct_index_offset`.
5.  Read the `DistinctIndex` using that index